### PR TITLE
Improve ViewerTransformChanger and add AdaptedViewCenterTransformCreator

### PR DESCRIPTION
**Pull request by tischi**
I added more functionality to the navigate package.

You can see it in practice by running the amended `ViewTransformSetAndLogDemo`.

### CODE REVIEW

File: src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import net.imglib2.realtransform.AffineTransform3D;
+
+public class AdaptedViewCenterTransformCreator
+{
+	private final BdvHandle bdvHandle;
+	private final double[] xyz;
+	private final int t;
+
+	/**
+	 * Based on the current ViewerTransform, creates a new viewer transform with a new center position.
```
Comment (NicoKiaru) [inline src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java:13]: Maybe move this to the header and put the documentation like the other actions.

Also I think we were trying to implement Runnable.

File: src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		final double[] windowCentreInViewerInPixels = new double[ 3 ];
+		windowCentreInViewerInPixels[ 0 ] = bdvHandle.getViewerPanel().getDisplay().getWidth() / 2.0;
+		windowCentreInViewerInPixels[ 1 ] = bdvHandle.getViewerPanel().getDisplay().getHeight() / 2.0;
+
+		adaptedViewerTransform.translate( windowCentreInViewerInPixels );
+
+		return adaptedViewerTransform;
+	}
+
+	public AffineTransform3D getAdaptedCenterTransform()
```
Comment (NicoKiaru) [inline src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java:60]: I would be in favor of adding a Run function like the other actions. The run function would center the BdvHandle to the given coordinates. Also it would be nice if you can give an Array of BdvHandle in order to recenter a set of BdvHandle.

The AffineTransform could be returned with a getter method

File: src/main/java/sc/fiji/bdvpg/bdv/navigate/ViewerTransformChanger.java [hunk 3]
```diff
@@ -38,31 +39,53 @@
  *
  * See ViewTransformSetAndLogDemo for a usage example
```
Comment (NicoKiaru) [inline src/main/java/sc/fiji/bdvpg/bdv/navigate/ViewerTransformChanger.java]: Explain in the header why there's an animation duration, and how setting a negative (or 0?) value leads to instant displacement.

File: src/main/java/sc/fiji/bdvpg/scijava/command/bdv/ViewTransformatorCommand.java [hunk 4]
```diff
@@ -78,6 +78,6 @@ public void run() {
         affineTransform3D.rotate(1, rotateAroundY);
         affineTransform3D.rotate(2, rotateAroundZ);
 
-        new ViewTransformator(bdvh, affineTransform3D).run();
+        new ViewerTransformChanger(bdvh, affineTransform3D, true, 0).run();
```
Comment (NicoKiaru) [inline src/main/java/sc/fiji/bdvpg/scijava/command/bdv/ViewTransformatorCommand.java:81]: Do you think it can be useful to have another constructor wiht true and 0 being the default values ?

**Review by NicoKiaru — COMMENTED** [review]
HI @tischi , I've done a few comments. Let me know what you think.

File: src/main/java/sc/fiji/bdvpg/scijava/command/bdv/ViewTransformatorCommand.java (hunk 4 above)
Comment (tischi) [reply src/main/java/sc/fiji/bdvpg/scijava/command/bdv/ViewTransformatorCommand.java:81]: Yes, I will do this.

File: src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java (hunk 2 above)
Comment (tischi) [reply src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java:60]: OK, I see what you mean. 

But then this method also needs the `animationDurationTime` input parameter of the `ViewerTransformChanger`.
Feels a bit like not so good separation of concerns to me.
In this sense keeping methods that _create_ viewer transforms (there could be many) and that _apply_ the viewer transform (maybe we have only one, but there could be also more in the future) separate feels better to me. What do you think?

I think the only reason one would need just one class that does both would be for the UI, and there I think we anyway would need to create a `Command` that then would wrap two calls like this: `new TransformApplier( new TransformCreator.get() ).run()`, isn't it?

So, in other words, I would prefer to put directly executable actions into `Command`s and keep things more separate on the backend API side.

File: src/main/java/sc/fiji/bdvpg/bdv/navigate/ViewerTransformChanger.java (hunk 3 above)
Comment (tischi) [reply src/main/java/sc/fiji/bdvpg/bdv/navigate/ViewerTransformChanger.java]: ok!

File: src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java (hunk 1 above)
Comment (tischi) [reply src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java:13]: Docu: OK
Runnable: I don't think this should be runnable (see my answer below).

Comment (tischi) [reply src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java:13]: ...but we could make it `Callable` if that helps, returning an `AffineTransform`.

File: src/main/java/sc/fiji/bdvpg/scijava/command/bdv/ViewTransformatorCommand.java (hunk 4 above)
Comment (tischi) [reply src/main/java/sc/fiji/bdvpg/scijava/command/bdv/ViewTransformatorCommand.java:81]: But I think we should not put `concatenateTransform = true` as a default, because I was in fact surprised when I saw that it was concatenating. I would thus put no default there but require it in the constructor.
Ok?

**Comment by tischi** [comment]
@NicoKiaru 
I saw you reacted to one of my answers. Are you still thinking about the other comments or did you overlook them? No rush, just to know whether I should still wait... ;-)

**Comment by NicoKiaru** [comment]
Yep doing this slowly in parallel of many other stuff.

File: src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java (hunk 2 above)
Comment (NicoKiaru) [reply src/main/java/sc/fiji/bdvpg/bdv/navigate/AdaptedViewCenterTransformCreator.java:60]: Ok, then why not BdvHandleHelper (this class may already exists) which gives  a few useful static function (centerViewOn, rescale, etc.) and only one or few actions which apply the transform on one or several BdvHandles ?

**Comment by tischi** [comment]
@NicoKiaru 
I tried to change according to our discussion.

**Comment by NicoKiaru** [comment]
thanks!
